### DISCUSSION TITLE
ScrollArea Scroll Bar customization?

### DISCUSSION THREAD

**Discussion by newcomb-luke** [emilk/egui#1593] [Q&A]
I am currently trying to sort of "port" an existing GTK application to Rust/egui, and I want it to have a similar feel to the GTK application. Obviously GTK has been around forever so it would have advanced features. I really want to mimic the behavior of it's scrollbars where you can set them to fade in/out based on if the user moves the cursor, and I want to be able to let it become thinner and "frameless" when your cursor is not hovering over it and have it animate out when you hover over/click on it.

This is a bit of an odd question, but I would like to see if I can implement that. The only problem that I currently see is that the current ScrollArea implementation uses the Context::frameState() method which is private. I would assume that is just so that users can't mess with a bunch of internals. My actual question comes in two pieces:

If I wanted to make this, is forking egui my only option? Implementing it as a widget seems... not the greatest. Any other way of doing it may be nice.

If I did make this and had to fork egui, is that something that egui's scope would support? If it does happen, would a pull request be a good idea?

Thank you.

**Comment by emilk** [comment]
`ScrollArea` is not customizable enough for you to change the behavior without a fork and a PR.

Can you show me a video of the effect you are going for?

Currently egui:s scrollbars take up actual space (as opposed to e.g. iOS). If they would shrink/grow on hover I think we need to make that effect be only visuals, or else the content rect would also become wider/narrower and the content would move and re-layout on hover, which seems annoying.

**Reply by newcomb-luke** [reply]
I know that the shrink/grow effect would definitely need to be only a visual effect as to not cause chaos on the actual contents.

Here is the effect I was mentioning about it growing/shrinking or even disappearing all together:
![scroll_hover](https://user-images.githubusercontent.com/16246849/168166769-84c7d442-9d26-4c63-9f7e-1db65ea61fa6.png)

Here is also the related effect when you scroll down to the bottom of a scrollable area, and it brings up that overlay curve at the bottom:
![image](https://user-images.githubusercontent.com/16246849/168167009-e75594a4-d4a0-4bd1-967d-aca2afb3804d.png)

I think it would be nice to add these, it would probably be a customization where you can enable it and probably have it off by default.
